File: ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/HumanLivingCharacter.swift
import Foundation

// MARK: - Living character (how you live, not who your uncle is)
//
// Closed chips only. No family tree, no addresses, no SF-86. ARIA keeps this
// on-device so split-second replies (Dummy, training flavor, Lifestyle) can
// read who you are as a human without a model round-trip.

public enum LivingMovementPreference: String, Codable, CaseIterable, Sendable {
    case cardio
    case strength
    case mix

    public var title: String {
        switch self {
        case .cardio: return "Cardio"
        case .strength: return "Strength"
        case .mix: return "A mix"
        }
    }

    public var livingTag: String { "living:move:\(rawValue)" }
}

public enum LivingEatingRhythm: String, Codable, CaseIterable, Sendable {
    case light
    case moderate
    case heavy

    public var title: String {
        switch self {
        case .light: return "Light eater"
        case .moderate: return "Moderate"
        case .heavy: return "I eat a lot"
        }
    }

    public var livingTag: String { "living:volume:\(rawValue)" }
}

public enum LivingHobby: String, Codable, CaseIterable, Sendable {
    case outdoors
    case gym
    case cooking
    case making
    case music
    case reading
    case games
    case rest

    public var title: String {
        switch self {
        case .outdoors: return "Being outside"
        case .gym: return "The gym"
        case .cooking: return "Cooking"
        case .making: return "Making things"
        case .music: return "Music"
        case .reading: return "Reading"
        case .games: return "Games"
        case .rest: return "Resting at home"
        }
    }

    public var livingTag: String { "living:hobby:\(rawValue)" }
}

public extension QualityOfLifePersona {
    /// Compact tag vocabulary Dummy / PlanEngine / Habit paths can read locally.
    func livingTags() -> [String] {
        var tags: [String] = []
        if archetype != .unset {
            tags.append("living:archetype:\(archetype.rawValue)")
        }
        if let food = nutritionRelationship?.trimmingCharacters(in: .whitespacesAndNewlines),
           !food.isEmpty {
            tags.append("living:eat:\(food.lowercased())")
        }
        if let rhythm = eatingRhythm {
            tags.append(rhythm.livingTag)
        }
        if let move = movementPreference {
            tags.append(move.livingTag)
        }
        for hobby in (hobbies ?? []).prefix(4) {
            tags.append(hobby.livingTag)
        }
        if let hours = sleepNeedPreferenceHours {
            tags.append(String(format: "living:sleep_need:%.1f", hours))
        }
        return tags
    }

    /// One paragraph ARIA can speak without calling the model.
    func characterLine() -> String {
        var parts: [String] = []
        if archetype != .unset {
            parts.append("You live like a \(archetype.title.lowercased()).")
        }
        if let move = movementPreference, move != .mix {
            parts.append("You typically like \(move.title.lowercased()).")
        } else if movementPreference == .mix {
            parts.append("You like a mix of cardio and strength.")
        }
        let hobbyTitles = (hobbies ?? []).prefix(3).map { $0.title.lowercased() }
        if !hobbyTitles.isEmpty {
            parts.append("Free days: \(hobbyTitles.joined(separator: ", ")).")
        }
        if let food = nutritionRelationship?.trimmingCharacters(in: .whitespacesAndNewlines),
           !food.isEmpty {
            var foodBit = "Food is \(food)"
            if let rhythm = eatingRhythm {
                foodBit += ", \(rhythm.title.lowercased())"
            }
            parts.append(foodBit + ".")
        } else if let rhythm = eatingRhythm {
            parts.append("\(rhythm.title).")
        }
        if let hours = sleepNeedPreferenceHours {
            parts.append(String(format: "Sleep want about %.1f hours.", hours))
        }
        parts.append("I keep this on-device so I don't have to ask the model who you are.")
        return parts.joined(separator: " ")
    }

    /// Short flavor for a local training decision — never a second personality quiz.
    static func livingDecisionNote(from tags: [String]) -> String? {
        let move = tags.first { $0.hasPrefix("living:move:") }
            .map { String($0.dropFirst("living:move:".count)) }
        let hobbies = tags.compactMap { tag -> String? in
            guard tag.hasPrefix("living:hobby:") else { return nil }
            return LivingHobby(rawValue: String(tag.dropFirst("living:hobby:".count)))?.title.lowercased()
        }
        var bits: [String] = []
        if let move, move != "mix", let pref = LivingMovementPreference(rawValue: move) {
            bits.append("You like \(pref.title.lowercased())")
        }
        if let first = hobbies.first {
            if hobbies.count > 1 {
                bits.append("free days lean \(first) and \(hobbies[1])")
            } else {
                bits.append("free days lean \(first)")
            }
        }
        guard !bits.isEmpty else { return nil }
        return bits.joined(separator: "; ") + " — pulled from your local profile."
    }
}

public extension QualityOfLifeLivingStore {
    static func livingTags(defaults: UserDefaults = .standard) -> [String] {
        loadPersona(defaults: defaults).livingTags()
    }

    /// Chat that is asking who they are as a human — answer from the local
    /// persona, never mint a model biography.
    static func isCharacterQuestion(_ text: String) -> Bool {
        if isQuestion(text) { return false }
        let lower = text.lowercased()
        let phrases = [
            "who am i",
            "who i am",
            "who you think i am",
            "how do i live",
            "how i live",
            "what do you know about me",
            "what you know about me",
            "my lifestyle profile",
            "my character",
            "what are my hobbies",
            "how do i usually train",
            "do i like cardio",
            "do i like strength",
        ]
        return phrases.contains { lower.contains($0) }
    }

    static func characterLine(
        defaults: UserDefaults = .standard,
        variety: Int = 0
    ) -> String {
        let persona = loadPersona(defaults: defaults)
        let hasLiving = persona.archetype != .unset && persona.archetype != .balanced
            || persona.movementPreference != nil
            || !(persona.hobbies ?? []).isEmpty
            || persona.nutritionRelationship != nil
            || persona.eatingRhythm != nil
        if !hasLiving && persona.archetype == .balanced
            && persona.movementPreference == nil
            && (persona.hobbies ?? []).isEmpty
            && persona.nutritionRelationship == nil {
            let missing = [
                "I don't have a living profile yet. Open Lifestyle and tell me how you live — how you eat, how you like to move, what fills a free day. I keep that on-device.",
                "No living character on file. Lifestyle interview is how I learn the shape of your week, not your family tree.",
                "Tell me how you live in Lifestyle and I'll keep it local — hobbies, eating rhythm, cardio vs strength. No model trip for that.",
            ]
            let index = variety < 0 ? 0 : variety
            return missing[index % missing.count]
        }
        let line = persona.characterLine()
        let openers = ["", "Same profile Life uses. ", "On-device, not a model guess: "]
        let index = variety < 0 ? 0 : variety
        return openers[index % openers.count] + line
    }
}

File: backend/infra/lambda/aria_core/human_living_character.py
"""Living-character tags Dummy / plan / habit paths can read without a model.

Python port of ForgeCore's ``HumanLivingCharacter.swift`` (closed chips +
copy). Persistence stays on the client.
"""

from __future__ import annotations

MOVEMENT_TITLE = {"cardio": "Cardio", "strength": "Strength", "mix": "A mix"}
EATING_TITLE = {"light": "Light eater", "moderate": "Moderate", "heavy": "I eat a lot"}
HOBBY_TITLE = {
    "outdoors": "Being outside",
    "gym": "The gym",
    "cooking": "Cooking",
    "making": "Making things",
    "music": "Music",
    "reading": "Reading",
    "games": "Games",
    "rest": "Resting at home",
}

ARCHETYPE_TITLE = {
    "athlete": "Athlete",
    "builder": "Builder",
    "caretaker": "Caretaker",
    "explorer": "Explorer",
    "thinker": "Thinker",
    "balanced": "Balanced",
    "unset": "Unset",
}


def living_tags(
    *,
    archetype: str | None = None,
    nutrition_relationship: str | None = None,
    eating_rhythm: str | None = None,
    movement_preference: str | None = None,
    hobbies: list[str] | None = None,
    sleep_need_preference_hours: float | None = None,
) -> list[str]:
    tags: list[str] = []
    if archetype and archetype != "unset":
        tags.append(f"living:archetype:{archetype}")
    food = (nutrition_relationship or "").strip()
    if food:
        tags.append(f"living:eat:{food.lower()}")
    if eating_rhythm:
        tags.append(f"living:volume:{eating_rhythm}")
    if movement_preference:
        tags.append(f"living:move:{movement_preference}")
    for hobby in (hobbies or [])[:4]:
        tags.append(f"living:hobby:{hobby}")
    if sleep_need_preference_hours is not None:
        tags.append(f"living:sleep_need:{sleep_need_preference_hours:.1f}")
    return tags


def character_line(
    *,
    archetype: str | None = None,
    nutrition_relationship: str | None = None,
    eating_rhythm: str | None = None,
    movement_preference: str | None = None,
    hobbies: list[str] | None = None,
    sleep_need_preference_hours: float | None = None,
) -> str:
    parts: list[str] = []
    if archetype and archetype != "unset":
        title = ARCHETYPE_TITLE.get(archetype, archetype).lower()
        parts.append(f"You live like a {title}.")
    if movement_preference and movement_preference != "mix":
        parts.append(f"You typically like {MOVEMENT_TITLE.get(movement_preference, movement_preference).lower()}.")
    elif movement_preference == "mix":
        parts.append("You like a mix of cardio and strength.")
    hobby_titles = [HOBBY_TITLE.get(h, h).lower() for h in (hobbies or [])[:3]]
    if hobby_titles:
        parts.append("Free days: " + ", ".join(hobby_titles) + ".")
    food = (nutrition_relationship or "").strip()
    if food:
        food_bit = f"Food is {food}"
        if eating_rhythm:
            food_bit += f", {EATING_TITLE.get(eating_rhythm, eating_rhythm).lower()}"
        parts.append(food_bit + ".")
    elif eating_rhythm:
        parts.append(EATING_TITLE.get(eating_rhythm, eating_rhythm) + ".")
    if sleep_need_preference_hours is not None:
        parts.append(f"Sleep want about {sleep_need_preference_hours:.1f} hours.")
    parts.append("I keep this on-device so I don't have to ask the model who you are.")
    return " ".join(parts)


def living_decision_note(tags: list[str]) -> str | None:
    move = None
    for tag in tags:
        if tag.startswith("living:move:"):
            move = tag[len("living:move:"):]
            break
    hobbies = []
    for tag in tags:
        if tag.startswith("living:hobby:"):
            raw = tag[len("living:hobby:"):]
            hobbies.append(HOBBY_TITLE.get(raw, raw).lower())
    bits: list[str] = []
    if move and move != "mix" and move in MOVEMENT_TITLE:
        bits.append(f"You like {MOVEMENT_TITLE[move].lower()}")
    if hobbies:
        if len(hobbies) > 1:
            bits.append(f"free days lean {hobbies[0]} and {hobbies[1]}")
        else:
            bits.append(f"free days lean {hobbies[0]}")
    if not bits:
        return None
    return "; ".join(bits) + " — pulled from your local profile."
